src/datastore.rs

```diff
@@ -94,59 +97,101 @@ pub struct MailboxPublicKey {
 #[sqlx(type_name = "key_type", rename_all = "lowercase")]
-enum KeyType { Secp256k1 }
+pub enum KeyType { Secp256k1 }
 
-impl sqlx::FromRow<'_, sqlx::postgres::PgRow> for PublicKeyResponse {
-    fn from_row(row: &sqlx::postgres::PgRow) -> std::result::Result<Self, sqlx::Error> {
-        let key_bytes: Vec<u8> = base64::decode(row.try_get::<&str, _>("public_key")?)
-            .map_err(|err| sqlx::Error::Decode(Box::new(err)))?;
-        let public_key = match row.try_get::<KeyType, _>("public_key_type")? {
-            KeyType::Secp256k1 => Key::Secp256k1(key_bytes),
-        };
-        let p8e_key_bytes: Vec<u8> = base64::decode(row.try_get::<&str, _>("signing_public_key")?)
-            .map_err(|err| sqlx::Error::Decode(Box::new(err)))?;
-        let signing_public_key = match row.try_get::<KeyType, _>("signing_public_key_type")? {
-            KeyType::Secp256k1 => Key::Secp256k1(p8e_key_bytes),
+#[derive(Clone, Debug)]
+pub struct PublicKey {
+    pub uuid: uuid::Uuid,
+    pub public_key: String,
+    pub public_key_type: KeyType,
+    pub url: String,
+    pub metadata: Vec<u8>,
+    pub auth_type: Option<AuthType>,
+    pub auth_data: Option<String>,
+    pub created_at: DateTime<Utc>,
+    pub updated_at: DateTime<Utc>,
+}
+
+impl PublicKey {
+    pub fn auth(&self) -> Result<Box<dyn Authorization + '_>> {
+        match self.auth_type {
+            Some(AuthType::Header) => {
+                let auth_data = self.auth_data.as_ref()
+                    .ok_or(OsError::InvalidApplicationState(String::from("auth_type was set but no auth_data")))?;
+                let (header, value) = auth_data.split_once(":")
+                    .ok_or(OsError::InvalidApplicationState(String::from("auth_data invalid format")))?;
+
+                Ok(Box::new(HeaderAuth { header, value }))
+            },
+            None => Ok(Box::new(NoAuthorization::default())),
+        }
+    }
+}
+
+impl TryFrom<PublicKeyRequest> for PublicKey {
+    type Error = OsError;
+
+    fn try_from(request: PublicKeyRequest) -> Result<Self> {
+        let (public_key_type, public_key) = match request.public_key.unwrap().key.unwrap() {
+            Key::Secp256k1(data) => (KeyType::Secp256k1, base64::encode(data)),
         };
-        let created_at: SystemTime = row.try_get::<DateTime<Utc>, _>("created_at")?.into();
-        let updated_at: SystemTime = row.try_get::<DateTime<Utc>, _>("updated_at")?.into();
-        let metadata: Vec<u8> = row.try_get("metadata")?;
-        let metadata = if !metadata.is_empty() {
-            let message = prost_types::Any::decode(metadata.as_slice())
-                .map_err(|err| sqlx::Error::Decode(Box::new(err)))?;
-            Some(message)
+        let metadata = if let Some(metadata) = request.metadata {
+            let mut buffer = BytesMut::with_capacity(metadata.encoded_len());
+            metadata.encode(&mut buffer)?;
+            buffer
         } else {
-            None
+            BytesMut::default()
```

Comment:
I think it's possible to use `Default::default()` universally anywhere you'd use `BytesMut::default()`, `Vec::default()`, etc. I don't necessarily know if that's a good practice though

Reply:
You're probably right. It has enough type information present to know. I do like explicitly using the concrete type though.